src/bin/miri.rs

```diff
@@ -63,6 +65,32 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
     }
 }
 
+struct OutputMirHack {
+    target_crate: bool,
+}
+
+impl rustc_driver::Callbacks for OutputMirHack {
+    fn config(&mut self, config: &mut Config) {
+        if config.opts.prints.is_empty() && self.target_crate {
+            // This is a hack to make
+            // https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L167
+            // always false.
+            config.opts.output_types = OutputTypes::new(
+                &config.opts.output_types.keys()
+                    .copied()
+                    .zip(config.opts.output_types.values().cloned())
+                    .chain(iter::once((
+                        OutputType::Mir,
+                        Some(if cfg!(unix) { "/dev/null" } else { "NUL" }.into()),
+                    )))
+                    .collect::<Vec<_>>(),
+            );
```

Comment:
A `rustc` patch seems necessary to get rid of this unless there's already a way force `rustc` to dump all exported symbols to Miri that I don't know about. Perhaps I could add a new `-Z` flag or let it dump them when `-Zalways-encode-mir` is set.

Reply:
So, what problem is this solving? Can't we just copy-paste [these 5 lines of code](https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L171-L175) instead?

Reply:
I didn't realize it's just 5 lines...

Reply:
Oh it may not work: https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L48-L50
I'm not sure. I will try that.

Reply:
Ah, that's annoying... why is the same thing checked *again* here?

Reply:
> Oh it might not work: https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L48-L50
> I'm not sure. I will try that.

I tried it and indeed it does not work. :/
The checks seems coming from rust-lang/rust#48611.

Reply:
Hm, that's a bummer.

I wonder if it would make sense to have our own version of [this](https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L62). We anyway care about slightly different things, namely only the `no_mangle` and `link_name` functions.

Reply:
I tried to do so but somehow broke `__rust_start_panic` when `panic=abort`. :thinking:
```
thread 'main' panicked at 'panicking from libstd', tests/compile-fail/panic/panic_abort1.rs:5:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: unsupported operation: multiple definitions of `__rust_start_panic` (in crate `panic_abort` and `panic_unwind`)
   --> /usr/share/rust/.rustup/toolchains/master/lib/rustlib/src/rust/library/std/src/panicking.rs:640:9
    |
640 |         __rust_start_panic(obj)
    |         ^^^^^^^^^^^^^^^^^^^^^^^ multiple definitions of `__rust_start_panic` (in crate `panic_abort` and `panic_unwind`)
    |
    = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
```
I also tried to change `SymbolExportLevel::C` to `SymbolExportLevel::Rust` but the result was same.

Reply:
Hm.... I guess there is some magic going to to only make one of `panic_abort` and `panic_unwind` actually appear -- and we probably do *not* want to reimplement that magic.

Reply:
I restored the `__rust_start_panic` and `__rust_panic_cleanup` forwarding code *for now* although I really want to remove them. :(

Reply:
I restored `config.opts.crate_types = vec![CrateType::Rlib]` and it started to work. I *guess* it's because `std` has `crate-type = ["dylib", "rlib"]`, and `rustc` thinks it must pull in `panic_unwind` to generate a `dylib`, not because `rustc` is a wizard.